offload/DeviceRTL/src/Workshare.cpp

```diff
@@ -444,32 +444,81 @@ template <typename T, typename ST> struct omptarget_nvptx_LoopSupport {
 // KMP interface implementation (dyn loops)
 ////////////////////////////////////////////////////////////////////////////////
 
-// TODO: This is a stopgap. We probably want to expand the dispatch API to take
-//       an DST pointer which can then be allocated properly without malloc.
-static DynamicScheduleTracker *THREAD_LOCAL(ThreadDSTPtr);
+// TODO: Expand the dispatch API to take a DST pointer which can then be
+//       allocated properly without malloc.
+// For now, each team will contain an LDS pointer (ThreadDST) to a global array
+// of references to the DST structs allocated (in global memory) for each thread
+// in the team. The global memory array is allocated during the init phase if it
+// was not allocated already and will be deallocated when the dispatch phase
+// ends:
+//
+//  __kmpc_dispatch_init
+//
+//  ** Dispatch loop **
+//
+//  __kmpc_dispatch_deinit
+//
+static DynamicScheduleTracker **SHARED(ThreadDST);
 
 // Create a new DST, link the current one, and define the new as current.
 static DynamicScheduleTracker *pushDST() {
+  int32_t ThreadIndex = mapping::getThreadIdInBlock();
+  // Each block will allocate an array of pointers to DST structs. The array is
+  // equal in length to the number of threads in that block.
+  if (!ThreadDST) {
+    // Allocate global memory array of pointers to DST structs:
+    if (ThreadIndex == 0)
```

Comment:
How about generic mode? The main thread is not the first thread in the block.

Reply:
Thread 0 is an arbitrary choice.

Reply:
This doesn't sound right. IIUC, in generic mode, this function needs to be called by the main thread before waking up worker threads.

Reply:
Is there a better check to do then?

Reply:
I'll check that in a bit, it looks like trunk is broken and I've rebased on latest. Trying to resolve the multiple fails with trunk to be able to actually build the compiler again.

Reply:
workshare loops have to be hit by all threads in the parallel region.
So if we are in an actual parallel, this should work, otherwise it is a single thread that is not thread 0. I assume you need to check for that (`isMainThreadInGenericMode`).

Reply:
Can we call this code and not be in a parallel?